x/staking/keeper/delegation.go

```diff
@@ -60,19 +60,70 @@ func (k Keeper) GetAllDelegations(ctx sdk.Context) (delegations []types.Delegati
 func (k Keeper) GetValidatorDelegations(ctx sdk.Context, valAddr sdk.ValAddress) (delegations []types.Delegation) {
 	store := ctx.KVStore(k.storeKey)
 
-	iterator := storetypes.KVStorePrefixIterator(store, types.DelegationKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.GetDelegationsByValPrefixKey(valAddr))
 	defer iterator.Close()
 
 	for ; iterator.Valid(); iterator.Next() {
-		delegation := types.MustUnmarshalDelegation(k.cdc, iterator.Value())
-		if delegation.GetValidatorAddr().Equals(valAddr) {
-			delegations = append(delegations, delegation)
+		var delegation types.Delegation
+		valAddr, delAddr, err := types.ParseDelegationsByValKey(iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+
+		bz := store.Get(types.GetDelegationKey(delAddr, valAddr))
+		if err := k.cdc.Unmarshal(bz, &delegation); err != nil {
+			panic(err)
 		}
+
+		delegations = append(delegations, delegation)
 	}
 
 	return delegations
 }
 
+func (k Keeper) UpdateValidatorDelegations(ctx sdk.Context, existingValAddr, newValAddr sdk.ValAddress) {
+	store := ctx.KVStore(k.storeKey)
+
+	itr := storetypes.KVStorePrefixIterator(store, types.GetDelegationsByValPrefixKey(existingValAddr))
+	defer itr.Close()
+
+	for ; itr.Valid(); itr.Next() {
+		key := itr.Key()
+		valAddr, delAddr, err := types.ParseDelegationsByValKey(key)
+		if err != nil {
+			panic(err)
+		}
+
+		bz := store.Get(types.GetDelegationKey(delAddr, valAddr))
+		delegation := types.MustUnmarshalDelegation(k.cdc, bz)
+
+		// remove old operator addr from delegation
+		if err := k.RemoveDelegation(ctx, delegation); err != nil {
+			panic(err)
+		}
+
+		delegation.ValidatorAddress = newValAddr.String()
+		// add with new operator addr
+		k.SetDelegation(ctx, delegation)
+	}
+}
+
+func (k Keeper) UpdateValidatorDelegationsLegacy(ctx sdk.Context, existingValAddr, newValAddr sdk.ValAddress) {
```

Comment:
ditto, When and where is this method called? Is it only in benchmarks?

Reply:
I don't think it's a good to define methods on keepers for the sole use of tests and/or benchmarks. Now we have an additional public API we need to support. Let's figure out a way to move this to the tests/benchmark logic.

Reply:
thanks for pointing out, moved to tests file.